client/src/Try/Gist.purs

```diff
@@ -5,45 +5,78 @@ module Try.Gist
 import Data.Either (Either(..))
-import Effect (Effect)
-import Effect.Uncurried (EffectFn1, EffectFn3, EffectFn4, mkEffectFn1, runEffectFn3, runEffectFn4)
+import Data.Function.Uncurried (Fn2, runFn2)
+import Data.Maybe (Maybe(..))
+import Data.Nullable (Nullable, toMaybe)
+import Effect.Aff (Aff)
+import Foreign.Generic (encodeJSON)
+import Foreign.Object as Object
+import Unsafe.Coerce (unsafeCoerce)
 
 -- | An abstract data type representing the data we get back from the GitHub API.
 data GistInfo
 
-foreign import uploadGist_
-  :: EffectFn3 String
-               (EffectFn1 String Unit)
-               (EffectFn1 String Unit)
-               Unit
-
--- | A wrapper for `uploadGist` which uses `ContT`.
-uploadGist :: String -> ExceptT String (ContT Unit Effect) String
-uploadGist content = ExceptT (ContT \k -> runEffectFn3 uploadGist_ content (mkEffectFn1 (k <<< Right)) (mkEffectFn1 (k <<< Left)))
-
--- | Get a gist by its ID
-foreign import getGistById_
-  :: EffectFn3 String
-               (EffectFn1 GistInfo Unit)
-               (EffectFn1 String Unit)
-               Unit
-
--- | A wrapper for `getGistById` which uses `ContT`.
-getGistById :: String -> ExceptT String (ContT Unit Effect) GistInfo
-getGistById id_ = ExceptT (ContT \k -> runEffectFn3 getGistById_ id_ (mkEffectFn1 (k <<< Right)) (mkEffectFn1 (k <<< Left)))
-
-foreign import tryLoadFileFromGist_
-  :: EffectFn4 GistInfo
-               String
-               (EffectFn1 String Unit)
-               (EffectFn1 String Unit)
-               Unit
-
-tryLoadFileFromGist :: GistInfo -> String -> ExceptT String (ContT Unit Effect) String
-tryLoadFileFromGist gi filename = ExceptT (ContT \k -> runEffectFn4 tryLoadFileFromGist_ gi filename (mkEffectFn1 (k <<< Right)) (mkEffectFn1 (k <<< Left)))
+type FileName = String
+
+type RawURL = String
+
+foreign import rawUrl_ :: Fn2 GistInfo FileName (Nullable RawURL)
+
+-- | Retrieve the URL for the raw contents of a particular file within a gist,
+-- | if that file exists as part of the gist.
+rawUrl :: GistInfo -> FileName -> Either String RawURL
+rawUrl gist filename = case toMaybe $ runFn2 rawUrl_ gist filename of
+  Nothing ->
+    Left $ "Gist does not contain a file named " <> filename
+  Just url ->
+    Right url
+
+uploadGist :: String -> ExceptT String Aff String
+uploadGist content = ExceptT $ AX.post AXRF.json "https://api.github.com/gists" requestBody >>= case _ of
+  Left e ->
+    pure $ Left $ "Unable to load Gist metadata: \n" <> printError e
+  Right { status } | status >= StatusCode 400 ->
+    pure $ Left $ "Received error status code: " <> show status
+  Right { body } ->
+    pure $ body # caseJsonObject (Left $ "Expected object in uploadGist, received: " <> stringify body) \obj ->
+      case Object.lookup "id" obj of
+        Just v | Just v' <- toString v -> Right v'
+        Nothing -> Left "No id key found."
+        _ -> Left "Key id was not a string."
```

Comment:
This seemed like the simplest way to pull out the `id` key from an object in the response without resorting to the FFI and without writing a new decoder. Like the `rawUrl_`  FFI code I think ideally we'd move to something like `codec-argonaut` in the future and replace this with a proper representation of the response from the API.